File: java-example/src/main/java/no/javatec/pubsubemulator/kotest/spock/service/ControllerService.java
package no.javatec.pubsubemulator.kotest.spock.service;

import no.javatec.pubsubemulator.kotest.spock.dto.SampleInputMessage;
import no.javatec.pubsubemulator.kotest.spock.dto.SampleReturnMessage;
import no.javatec.pubsubemulator.kotest.spock.pubsub.DemoPublisher;
import jakarta.inject.Singleton;
import lombok.RequiredArgsConstructor;

/**
 * Service for emulating post request processing
 */
@Singleton
@RequiredArgsConstructor
public class ControllerService {

    private final DemoPublisher demoPublisher;

    public SampleReturnMessage processRequest(SampleInputMessage inputMessage) {
        final var returnMessage =
                new SampleReturnMessage("Hello " + inputMessage.getName() + ", thank you for sending the message");

        demoPublisher.send(returnMessage);
        return returnMessage;
    }
}

File: kotlin-example/src/main/kotlin/no/javatec/pubsubemulator/kotest/service/ControllerService.kt
package no.javatec.pubsubemulator.kotest.service

import jakarta.inject.Singleton
import no.javatec.pubsubemulator.kotest.dto.SampleInputMessage
import no.javatec.pubsubemulator.kotest.dto.SampleReturnMessage
import no.javatec.pubsubemulator.kotest.pubsub.DemoPublisher

@Singleton
class ControllerService(private val demoPublisher: DemoPublisher) {

    fun processRequest(inputMessage: SampleInputMessage): SampleReturnMessage =
        SampleReturnMessage("Hello ${inputMessage.name}, thank you for sending the message").also {
            demoPublisher.send(it)
        }
}